### DISCUSSION TITLE
Integration of Laravel Reverb with Flutter Mobile App WebSocket

### DISCUSSION THREAD

**Discussion by najeeb-anwari** [laravel/framework#51766] [Q&A] [answered]
Description:
I'm currently working on a project where I've implemented Laravel Reverb for WebSocket communication on the backend. Now, I'm in the process of developing a mobile app using Flutter for the client side. I'm seeking advice on the best approach to integrate the WebSocket functionality from Laravel Reverb into my Flutter mobile app.

Specifically, I'd like guidance on:

Setting up WebSocket connection handling in Flutter.
Establishing communication between the Flutter app and the Laravel backend via WebSocket.
Handling events and data transmission between the Flutter client and Laravel server effectively.
Any potential pitfalls or best practices to keep in mind during the integration process.
Any insights, code snippets, or recommended libraries would be greatly appreciated. Thank you in advance for your assistance!

**Comment by mrbohem** [comment]
Have you find any solution

**Reply by najeeb-anwari** [reply]
Not yet.

**Comment by n4ff4h** [comment]
I use this https://pub.dev/packages/dart_pusher_channels. The only one maintained well so far according to my research. Couldn't get to work with private channels though. Public channels does work however.

**Reply by najeeb-anwari** [reply]
Oh really!
Could you share the solution here. Your help is appreciated.

**Reply by juliansteenbakker** [reply]
I tried this package, however with private channels, even when adding the authorization header, i still receive null on the `$user` in:
```php
Broadcast::channel('App.Models.User.{id}', function ($user, $id) {
    return (int) $user->id === (int) $id;
});
```

How did you solve this @oolayemi ?

**Comment by mkarimshahbaz** [comment, marked as answer]
You can use web_socket_channel.

**Reply by najeeb-anwari** [reply]
Thank you very much. That really works. You saved my life.

**Reply by eslam-eng** [reply]
Could you share the solution here. Your help is appreciated

**Reply by EvgenyDavydenko** [reply]
could you show an example of the source code for connecting to a private channel? I can't pass the authorization token in such a way that I get a key like

``` "auth": "d6afo3fjzrmflyhq6mrg:418a1159a7b726e809ff9300a9b9440fd70109ee979a59ef9e9c3fbfdd5e9fc9" ```

**Reply by musibaforreal** [reply]
@EvgenyDavydenko did you have any luck with finding the solution

**Reply by newQuery** [reply]
@najeeb-anwari were you able to connect to private channels ?

**Comment by aim-ademola** [comment]
please i need help on this also

**Comment by EzekielWachira** [comment]
What about on android, what's the best library I can use?

**Reply by Cxatek18** [reply]
Have you found the answer?

**Comment by jbeduya** [comment]
I also have this problem for a long time. This is how I was able to make it work using `dart_pusher_channels`.

I added the following in my `bootstrap/app.php` ([source](https://laracasts.com/discuss/channels/laravel/laravel-reverb))

```php
    ...
    ->withBroadcasting(
        channels: __DIR__ . '/../routes/channels.php',
        attributes: ['prefix' => 'api', 'middleware' => ['auth:sanctum']],
    )
   ...
```

Then did this in flutter.

```dart
import 'dart:async';

import 'package:dart_pusher_channels/dart_pusher_channels.dart';

Future<void> connect() async {
  const hostOptions = PusherChannelsOptions.fromHost(
    scheme: 'ws',
    host: '192.168.1.2',
    key: '<YOUR REVERB KEY>',
    shouldSupplyMetadataQueries: true,
    metadata: PusherChannelsOptionsMetadata.byDefault(),
    port: 8080,
  );

  final client = PusherChannelsClient.websocket(
      options: hostOptions,
      connectionErrorHandler: (exception, trace, refresh) {
        refresh();
      });

  final channel = client.publicChannel(
    'user.5.wallet',
  );

  final private = client.privateChannel(
    'private-App.Models.User.5',
    authorizationDelegate:
        EndpointAuthorizableChannelTokenAuthorizationDelegate.forPrivateChannel(
      authorizationEndpoint: Uri.parse('http://192.168.1.2:8000/api/broadcasting/auth'),
      headers: const {
        'Authorization': 'Bearer <USER SANCTUM TOKEN>',
      },
    ),
  );

  StreamSubscription<ChannelReadEvent> channelSubscription =
      channel.bind('App\\Events\\WalletUpdatedEvent').listen((event) {
    print('Event received: ${event.data}');
  });

  StreamSubscription<ChannelReadEvent> privateSubscription =
      private.bind('App\\Events\\WalletUpdatedEvent').listen((event) {
    print('Private event received: ${event.data}');
  });

  client.onConnectionEstablished.listen((s) {
    print('Connection established');
    channel.subscribe();
    private.subscribe();
  });

  await client.connect();
}
```

**Reply by sarmaRandeepGithub12345** [reply]
> ```dart
> Connection established
> ```

hi i am still not able to connect despite writing the above code? what to do?

**Reply by jbeduya** [reply]
Do you have events that fire that you can listen to?

**Reply by sarmaRandeepGithub12345** [reply]
> Do you have events that fire that you can listen to?
I tried to print the error ,this  is what i get:
I/flutter (23614): error of socket is : SocketException: Connection timed out (OS Error: Connection timed out, errno = 110), address = 171.48.95.81, port = 44880

my message event
`
class MessageSent implements ShouldBroadcastNow
{
    use Dispatchable, InteractsWithSockets, SerializesModels;
    public $message;
    public function __construct($message)
    {
        $this->message = $message;
    }

    /**
     * Get the channels the event should broadcast on.
     *
     * @return array<int, \Illuminate\Broadcasting\Channel>
     */
    public function broadcastOn(): array
    {
        $channels = [];

        $participants = ChatParticipants::where('chat_id', $this->message->chat_id)->pluck('user_id');

        foreach ($participants as $userId) {
            $channels[] = new PrivateChannel('chat-channel.' . $userId);
        }
        return $channels;
    }
    // public function broadcastAs(): string
    // {
    //     return 'message.sent';
    // }
}`

**Reply by sarmaRandeepGithub12345** [reply]
> Do you have events that fire that you can listen to?
changes I have made in bootstrap/app.php

`return Application::configure(basePath: dirname(__DIR__))
    ->withRouting(
        web: __DIR__.'/../routes/web.php',
        api: __DIR__.'/../routes/api.php',
        commands: __DIR__.'/../routes/console.php',
        // channels: __DIR__.'/../routes/channels.php',
        health: '/up',
    )
    ->withBroadcasting(
        channels: __DIR__.'/../routes/channels.php',
        attributes: ['prefix' => 'api', 'middleware' => ['auth:sanctum']],
    )
    ->withMiddleware(function (Middleware $middleware) {
        //
    })
    ->withExceptions(function (Exceptions $exceptions) {
        //
    })->create();

`
changes in routes/channels.php for authentication
`
Broadcast::channel('chat-channel.{id}', function (User $user, $id) {
    return $user->id === $id;
});`
code to fire the event 
`broadcast(new MessageSent($message))->toOthers();`

**Reply by sarmaRandeepGithub12345** [reply]
> Do you have events that fire that you can listen to?

my dart code to connect to reverb and listen to events :
`
    Future<void> dartconnect() async {
        try {
          final auth = Provider.of<AuthState>(context, listen: false);
          const hostOptions = PusherChannelsOptions.fromHost(
            scheme: 'ws',
            host: '171.48.95.80',
            key: 'reverb_key',
            shouldSupplyMetadataQueries: true,
            metadata: PusherChannelsOptionsMetadata.byDefault(),
            port: 8080,
          );
    
          final client = PusherChannelsClient.websocket(
              options: hostOptions,
              connectionErrorHandler: (exception, trace, refresh) {
                refresh();
              });
    
          final private = client.privateChannel(
            'chat-channel.${auth.user!.id}',
            authorizationDelegate:
                EndpointAuthorizableChannelTokenAuthorizationDelegate
                    .forPrivateChannel(
              authorizationEndpoint: Uri.parse('$domain/api/broadcasting/auth'),
              headers: {
                'Authorization': 'Bearer ${auth.loginToken}',
                'Content-Type': 'application/json',
              },
            ),
          );
    
              StreamSubscription<ChannelReadEvent> privateSubscription =
                  private.bind('App\\Events\\MessageSent').listen((event) {
                print('Private event received:${event.data} ');
              });
        
              client.onConnectionEstablished.listen((s) {
                print('Connection established');
                private.subscribe();
              });
              print("just before request");
              await client.connect();
              print("just after request");
            } catch (e) {
              print("Datetime ${DateTime.now()} error $e");
            }
          }

 `

**Reply by sarmaRandeepGithub12345** [reply]
> Do you have events that fire that you can listen to?

Bro please help me🙏

**Reply by musibaforreal** [reply]
@jbeduya what form authorization were you using for the private channels were you using laravel sanctum tokens or any other forms of auth

**Reply by musibaforreal** [reply]
@sarmaRandeepGithub12345 please check your event names as they are case sensitive please make sure the name of the event in laravel match with the flutter one exactly

**Comment by jbeduya** [comment]
I am using Sanctum.

**Reply by musibaforreal** [reply]
would you mind sharing your project as i am having a problem with the auth for connecting with private channels always sending back a 403 error code when truing to connect

**Reply by musibaforreal** [reply]
i found the solution the channels are case sensitive so if your channel is User and tried to connect using user it will not work

**Comment by reach-211** [comment]
It's kinda late here, but I wrote a blog post that describes how you can actually accomplish this: https://blog.niy.ai/2025/01/28/integrating-flutter-with-laravel-reverb-a-practical-guide/

**Comment by amargiovanni** [comment]
if it can be helpful: https://github.com/Oltrematica/flutter_reverb

**Comment by alyAly12** [comment]
guy i face an issue here is my chatdata source it connexted to the web socket but it doesnt listen or stream ''any ideas''
''class NewChatRemoteDataSource {
  final ApiService apiService;
  late WebSocketChannel channel;
  bool _isDisposed = false;
  late PusherChannelsClient _pusher;
  final String token;
  Timer? _keepAliveTimer;
  final _connectionController = StreamController<bool>.broadcast();

  NewChatRemoteDataSource({
    required this.apiService,
    required this.token,
  }) {
    _initWebSocket();
  }

  Stream<bool> get connectionStatus => _connectionController.stream;


  void _initWebSocket() {
    print('Initializing WebSocket connection...');
    _connect();

    _keepAliveTimer = Timer.periodic(
      const Duration(seconds: 25),
          (_) => _sendKeepAlive(),
    );
  }

  void _connect() {
    channel = IOWebSocketChannel.connect(
      Uri.parse('wss://stg-api.sellity.sa/app/local-app-key?protocol=7&client=js&version=8.4.0&flash=false'),
      headers: {'Authorization': 'Bearer $token'},
    );

    channel.stream.listen(
          (data) => _handleSocketData(data),
      onError: (error) => _handleSocketError(error),
      onDone: () => _handleSocketDisconnect(),
    );
  }

  void _handleSocketData(dynamic data) {
    try {
      final jsonData = json.decode(data);
      if (jsonData['event'] == 'pusher:connection_established') {
        _connectionController.add(true);
        print('WebSocket connection established');
      }
    } catch (e) {
      print('Error parsing connection confirmation: $e');
    }
  }

  void _handleSocketError(dynamic error) {
    print('WebSocket error: $error');
    _connectionController.add(false);
    if (!_isDisposed) _reconnect();
  }

  void _handleSocketDisconnect() {
    print('WebSocket connection closed');
    _connectionController.add(false);
    if (!_isDisposed) _reconnect();
  }

  void _reconnect() {
    Future.delayed(const Duration(seconds: 3), () {
      if (!_isDisposed) {
        print('Attempting reconnect...');
        _connect();
      }
    });
  }

  void _sendKeepAlive() {
    if (!_isDisposed && channel.closeCode == null) {
      channel.sink.add(jsonEncode({'event': 'pusher:ping'}));
    }
  }

  Stream<ChatMessageModel> subscribeToMessages(int chatId) {
    final channelName = 'chat.$chatId';

    // Send subscription request
    channel.sink.add(jsonEncode({
      'event': 'pusher:subscribe',
      'data': {
        'channel': channelName,
        'auth': 'Bearer $token',
      }
    }));

    return channel.stream.asyncExpand((event) async* {
      try {
        final jsonData = json.decode(event);

        if (jsonData['event'] == 'pusher_internal:subscription_succeeded') {
          print('Subscribed to $channelName');
          return;
        }

        if (jsonData['event'] == 'ChatMessageSent' &&
            jsonData['channel'] == channelName) {
          final dynamic data = jsonData['data'];

          final parsedData = data is String ? json.decode(data) : data;
          yield ChatMessageModel.fromJson(parsedData);
        }
      } catch (e, stack) {
        print('Error processing message: $e\n$stack');
      }
    });
  }

  void sendRealTimeMessage(ChatMessageModel message) {
    final payload = {
      'event': 'ChatMessageSent',
      'channel': 'chat.${message.chatId}',
      'data': message.toJson()
    };

    if (!_isDisposed && channel.closeCode == null) {
      channel.sink.add(jsonEncode(payload));
      print('Sent real-time message: ${message.content}');
    }
  }

  Future<void> dispose() async {
    _isDisposed = true;
    _keepAliveTimer?.cancel();
    await channel.sink.close();
    await _connectionController.close();
  }


  // Rest of your existing methods...
  Future<List<ChatMessageModel>> fetchMessages(int chatId) async {
    final response = await apiService.get(
        path:
        ApiConstant.baseUrl + ApiConstant.retrieveMessagesEndToEnd(chatId));
    final List<dynamic> data = response.data['data'];
    return data.map((json) => ChatMessageModel.fromJson(json)).toList();
  }

  Future<ChatMessageModel> sendMessage(int chatId, String text, String type) async {
    final response = await apiService.post(
      path: ApiConstant.baseUrl + ApiConstant.sendChatMessage(chatId),
      data: {'content': text, 'type': type},
    );
    final message = ChatMessageModel.fromJson(response.data['data']);
    sendRealTimeMessage(message); // Sync with WebSocket
    return message;
  }

  Future<CreateNewChatModel> createChat(int productId) async {
    final response = await apiService.postWithoutData(
        path: ApiConstant.baseUrl + ApiConstant.createChat(productId));
    return CreateNewChatModel.fromJson(response.data['data']);
  }
}
